[pkg/collector] Add worker utilization stats to expvars

sgnn7 opened the pull request:
### What does this PR do?

This PR:
- Adds periodic (every 15s) collection of utilization stats (how much time was spent running checks) for each worker.
- Calculates an average over 5 minutes or however long the worker was running, whichever is less
- Updates the `expvars["runner"]["Workers"]["Instances"]["instance_name"].Utilization` expvar after each new data point
- Changes `expvars["runner"]["Workers"]` expvar that contained the worker count to `expvars["runner"]["Workers"]["Count"]`
- ~Shows a warning in log if any worker that is not running a long check (e.g. jmxfetch) utilization goes over a predefined utilization threshold (`0.99`)~ (removed pending follow-up task)

### Motivation

It's not easy to track down problems with worker utilization just from logs themselves. This change should allow much quicker debugging of busy agents.

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

#### Basic functionality

- Set some number of `check_runners` in config
- Run the agent
- Initially the worker stats (`curl -s localhost:5000/debug/vars | jq ".runner.Workers" `) show correctly:
   - Names of each runner and the utilization is `0`
   - Correct count of the runners
- After 15s and over time, the value will stabilize and ensure:
   - Names of each runner and the utilization (range: 0 < util < 1)
   - Correct count of the runners
- Stop the agent
- Change the `check_runners` to 1 and start the agent
-  Initially the worker stats (`curl -s localhost:5000/debug/vars | jq ".runner.Workers" `) show correctly:
   - Correct count (`1`) of the runners
   - Name of the runner (`worker_*`) and the utilization is `0`
- After 15s and over time, the value will stabilize and ensure:
   - Correct count of the runners
   - Name of the runner (`worker_*`) 
   - Utilization proportional to the multiple of the original test scenario (e.g.`0.02` avg per worker w/ `check_runners = 4` should be close to `0.08` utilization for 1 worker with a max of `1.00`)
   
#### Basic functionality pt2

- Using the values collected before, add more checks (or make them run longer) and observe the increase in utilization
- Using the values collected before, remove some checks (or make them run shorter) and observe the decrease in utilization
- Using the values collected before, disable/remove all checks and observe the utilization stabilize at 0 after 5 min
- A long-check (eg. jmxfetch check) should have a worker stuck at 1.0 utilization (as it's keeping the child process running).

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

sgnn7 committed:
[pkg/util] Add SlidingWindow util class

This object can be used to calculate a sliding window of polled values
over the configured period of time. Usage of this class will be needed
to implement the worker/runner utilization values to the expvars.

sgnn7 committed:
[pkg/collector/runner] Changed `running` expvar map to `Running`

This field capitalization was not consistent with other fields in the
expvars.

sgnn7 committed:
[pkg/collector/runner] Add worker instance info to published expvars

We now show the list of workers and their placeholder utilization
variables in the expvar list ("runner" -> "Workers" -> "Instances").

sgnn7 committed:
[pkg/collector/worker] Use custom string for worker's external presentation

Integer ID is not a good identifier for an instance of an object so we
now use `worker_<id>` for `worker.Workers`.

sgnn7 committed:
[pkg/collector/runner] Move worker expvar management from runner to worker

Since the runner shouldn't care about the internal management of the
stats for a worker, we now are doing that work in the worker directly.

sgnn7 committed:
[pkg/util] Make SlidingWindow use more ergonomic API

old code was prone to circular references as you would be creating a new
owner at the same time that you may be defining a closure for the
sliding window that will need the owner object address as well. By
separating creation and configuration, we now have a better setup for
non-circular references.

sgnn7 committed:
[pkg/collector/worker] Added code/tests to worker to call utilization tracker

Even though the utilization tracker is not in the codebase yet, this
change ensures that the expected function of this TBD component from the
worker's perspective behaves appropriately.

sgnn7 committed:
[pkg/collector/worker] Added UtilizationTracker class

This class can use the SlidingWindow and internal time tracking to
properly evaluate the utilization stats and to publish them to expvars.

sgnn7 committed:
[pkg/collector/worker] Fix linting issues in `utilization_tracker.go`

sgnn7 committed:
[pkg/collector/runner] Fix minor spelling error

This was blocking the build so we needed to fix it.

sgnn7 committed:
[pkg/collector] Fix golang-lint linting issues

These were preventing the build from succeeding so the issues have been
fixed now.

sgnn7 committed:
Merge remote-tracking branch 'origin/main' into sgnn7/runner-utilization

sgnn7 committed:
[pkg/collector] Fix flaky tests

Since we use timing-based tests, some build machines do not handle the
currently-set short frames well so they have been increased to
accomodate.

sgnn7 committed:
[pkg/collector] Make utilization of long-running checks be 100%

Since we want to count long-running tasks as 100%, we now count their
runtime in the stats. Some architectural changes needed to be made
because we wanted to ensure that we're not printing any warnings about
its value being 1.0 for that particular worker.

sgnn7 committed:
[pkg/collector] Fix more utilization timing tests

We needed even more tweaks to fix build issues in various environments.
While the tests take a bit longer to run now, they should be a bit less
frail.

sgnn7 committed:
[pkg/collector] Fix more async test race conditions

We had more places where slow execution could lead to either a panic
and/or assertion failures in tests so this commit fixes those.

sgnn7 commented:
Moving milestone to 7.32.0 since it look unlikely that this will get in before the freeze.

sgnn7 committed:
[pkg/collector/worker] Fix MacOS tests for `utilization_tracker.go`

It seems like MacOS is extremely inaccurate in timings so this PR adds
extra time to collect datapoints in the accuracy test.

sgnn7 committed:
[pkg/collector/worker] Fix MacOS tests for `utilization_tracker.go` pt2

It seems like MacOS is extremely inaccurate in timings so this PR adds
extra time to collect datapoints in the accuracy test (second attempt).

djmitche commented on pkg/collector/worker/utilization_tracker.go:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	if ut.stopped {
+		return fmt.Errorf("Attempted to use UtilizationTracker.Start() when the tracker after it was stopped")
```
```suggestion
		return fmt.Errorf("Attempted to use UtilizationTracker.Start() after the tracker was stopped")
```

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+// that should be included in the slidingWindow data
+type PollingFunc func() (value float64)
+
+// CallbackFunc is invoked when the polling function collects data and
+// we need to use the newly-calculated values
+type CallbackFunc func(SlidingWindow)
```
This comment is pretty confusing.  Is it called when the stats are updated?  Maybe `StatsUpdatedFunc`?

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+
+// SlidingWindow is the public API that we expose from the slidingWindow object
+type SlidingWindow interface {
+	Start(PollingFunc, CallbackFunc) error
+	Stop()
+
+	Average() float64
+	WindowSize() time.Duration
```
It'd be good to have a doc comment on each method in the interface (in any interface, I think..)

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+// and starts the ticker. We use `Start` to define most of the variables
+// instead of the constructor as we are likely to need the SlidingWindow
+// instance in the polling/callback closures, leading to a chicken/egg
+// problem in usage.
```
For `callbackFunc` the SlidingWindow is passed to the callback.  If the same were done for `pollingFunc`, would this issue be eliminated?

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+
+				sw.bucketsLock.Unlock()
+
+				sw.bucketIdx = (sw.bucketIdx + 1) % sw.numBuckets
```
This should probably be covered by bucketsLock?

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+type slidingWindow struct {
+	bucketIdx       int
+	buckets         []float64
+	bucketsLock     sync.RWMutex
```
what fields does this protect (similarly for windowLock)?

djmitche commented on pkg/util/sliding_window.go:
```diff
@@ -0,0 +1,218 @@
+// sliding window range.
+func (sw *slidingWindow) Average() float64 {
+	if !sw.isInitialized() {
+		log.Warnf("Attempting to use SlidingWindow.Average() without initializting it!")
```
```suggestion
		log.Warnf("Attempting to use SlidingWindow.Average() without initializing it!")
```

djmitche reviewed (commented):
A few comments below.  This generally looks good!

sgnn7 committed:
Update pkg/collector/worker/utilization_tracker.go

Co-authored-by: Dustin J. Mitchell <dustin.mitchell@datadoghq.com>

sgnn7 committed:
Update pkg/util/sliding_window.go

Co-authored-by: Dustin J. Mitchell <dustin.mitchell@datadoghq.com>

sgnn7 committed:
[pkg/util] Rename SlidingWindow's `CallbackFunc` to `StatsUpdateFunc`

Old name was not very obvious as to what it did or when it might be
invoked so this change ensures that the langue is a bit more clear.

sgnn7 committed:
[pkg/util] Add comments to `SlidingWindow` interface members

While the comments were present on the implementation, they were missing
from the interface so this commit adds those.

sgnn7 committed:
[pkg/collector/runner] Add comments to `UtilizationTracker` interface members

While the comments were present on the implementation, they were missing
from the interface so this commit adds those.

sgnn7 committed:
[pkg/util] Extend `SlidingWindow` bucketLock to include `bucketIdx` changes

While not strictly necessary, the original code was relying on assumptions
about the rest of the code in the object that may not hold true in the
future (about who is changing the `sw.bucketIdx` and `sw.numBuckets`
variables.

sgnn7 committed:
Merge remote-tracking branch 'origin/main' into sgnn7/runner-utilization

sgnn7 committed:
[pkg/collector/worker] Add looser timings to UtilizationTracker accuracy test

MacOS tests still seem to be extremely inaccurate so the delta expected
for the final check of this test has been increased.

sgnn7 committed:
[pkg/util] Clarify `SlidingWindow` lock naming

We now have some comments and better names for locks in this object. As
part of the changes, a test was cleaned up to be less likely to
cause deadlocks.

sgnn7 committed:
[pkg/util] Cleaned up `SlidingWindow` overlapping lock usage

Since the two sets of locks in `slidingWindow` were added at different
times, their usage wasn't as well coordinated as it needed to be. By
moving some code around and ensuring that we lock minimal sections of
the code, we now should have no overlap in the two lock footprints.

sgnn7 committed:
[pkg/util] Fix `sliding_window_test` changed lock type

This change reverts the mistaken RWLock -> Lock in the
`TestSlidingWindowAccuracy` test.

sgnn7 committed:
[tests] Try fixing MacOS CircleCI tests pt3

More timing issues fix attempts as we try to run this code on MacOS via
CircleCI.

sgnn7 committed:
[pkg/util] Ensure that SlidingWindow polling loop exits early if polling is stopped

Depending on how locks were used in the parent object, the old code was
liable to hit deadlocks in the sliding window since the polling function
could run during a Stop() event. This change ensures that we check if
we're stopped before invoking either of the callbacks.

djmitche reviewed (approved):
Looks good! 👍🏽
